src/fhda/tortoise.py

```diff
@@ -0,0 +1,256 @@
+        default_factory=dict, description="Arguments to format the prompt template"
+    )
+    input_files: Dict[str, str] = Field(
+        default_factory=dict, description="Files to upload {'source_path': 'dest_name'}"
+    )
+    output_files: Dict[str, str] = Field(
+        default_factory=dict,
+        description="Files to download {'source_name': 'dest_path'}",
+    )
+    step_id: str = Field(
+        default_factory=lambda: str(uuid.uuid4())[:8],
+        description="Small UID for the step",
+    )
+    upload_id: Optional[str] = Field(default=None, description="Upload ID for GCS")
+    parallel: int = Field(default=1, description="Number of parallel tasks to run")
+    config: StepConfig = Field(
+        default_factory=StepConfig, description="Configuration for the step"
+    )
+    post_process: Optional[Callable] = Field(
+        default=None, description="Function to run after step completion"
+    )
+    prompt_generator: Optional[Callable] = Field(
+        default=None,
+        description="Function to generate prompts and args for parallel tasks based on previous results",
+    )
+
+    def cot_prompting(self, query, language):
+        guidelines = prompts.GENERAL_NOTEBOOK_GUIDELINES.format(language=language)
+        if language == "R":
+            guidelines = prompts.R_SPECIFIC_GUIDELINES.format(language=language)
+        return (
+            f"{prompts.CHAIN_OF_THOUGHT_AGNOSTIC.format(language=language)}\n"
+            f"{guidelines}"
+            f"Here is the research question to address:\n"
+            f"<query>\n"
+            f"{query}\n"
+            f"</query>\n"
+        )
+
+    def format_prompt(self) -> str:
+        """Format the prompt template with the provided arguments"""
+        final_prompt = self.prompt.format(**self.prompt_args)
+        if self.cot_prompt:
+            final_prompt = self.cot_prompting(final_prompt, self.config.language)
+        return final_prompt
+
+
+class Tortoise:
+    """Framework for running multi-step agent pipelines"""
+
+    def __init__(self, api_key: str):
+        """Initialize the tortoise framework with FutureHouse API key"""
+        self.client = FutureHouseClient(auth_type=AuthType.API_KEY, api_key=api_key)
+        self.steps: List[Step] = []
+        self.results: Dict[str, Any] = {}
+
+    def add_step(self, step: Step) -> None:
+        """Add a step to the pipeline"""
+        self.steps.append(step)
+
+    def save_results(self, output_dir: str = "output") -> None:
+        """Save the results to a JSON file"""
+        results_path = f"{output_dir}/results_{time.strftime('%Y%m%d_%H%M%S')}.json"
+        print(f"Saving all results to {results_path}")
+        try:
+            os.makedirs(output_dir, exist_ok=True)
+            serializable_results = {}
+            for step_id, step_result in self.results.items():
+                serializable_step = dict(step_result)
+                serializable_results[step_id] = serializable_step
```

Comment:
```suggestion
                serializable_results[step_id] = dict(step_result)
```

We don't need `serializable_step` local, it's indirection